### ISSUE TITLE
Saving tf.keras.Sequential model fails with RNN containing more than one GRUCell

### ISSUE THREAD

**Issue by padoremu** [tensorflow/tensorflow#36093]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution: Linux Ubuntu 18.04
- Mobile device if the issue happens on mobile device: -
- TensorFlow installed from: binary (tf-nightly-2.0-preview)
- TensorFlow version: GIT_VERSION = v1.12.1-7529-g3e0ad8a004, VERSION = 2.0.0-dev20190731
- Python version: 3.6.9
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: CPU only
- GPU model and memory: CPU only

**Describe the current behavior**
Saving a `tf.keras.Sequential` model with `tf.keras.layers.RNN` fails when containing more than one `tf.keras.layers.GRUCell` with error message `RuntimeError: Unable to create link (name already exists)`.

**Describe the expected behavior**
Saving should succeed, not only when having one cell.

**Code to reproduce the issue**
```python
import tensorflow as tf

# saving succeeds for number_of_cells = 1, but fails for number_of_cells > 1
number_of_cells = 2

model = tf.keras.Sequential()

model.add(tf.keras.layers.Input(shape=(1, 1,)))

cells = []

for i in range(number_of_cells):
    cells.append(tf.keras.layers.GRUCell(10))

model.add(tf.keras.layers.RNN(cells))

model.save("rnn.h5")
```

**Other info / logs**
Behavior is the same when using `SimpleRNNCell` instead of `GRUCell`.

Traceback in case of failure:
```bash
Traceback (most recent call last):
  File "test_rnn_gru.py", line 17, in <module>
    model.save("rnn.h5")
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/network.py", line 1157, in save
    saving.save_model(self, filepath, overwrite, include_optimizer, save_format)
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/keras/saving/save.py", line 105, in save_model
    model, filepath, overwrite, include_optimizer)
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/keras/saving/hdf5_format.py", line 103, in save_model_to_hdf5
    save_weights_to_hdf5_group(model_weights_group, model_layers)
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/keras/saving/hdf5_format.py", line 625, in save_weights_to_hdf5_group
    param_dset = g.create_dataset(name, val.shape, dtype=val.dtype)
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/h5py/_hl/group.py", line 139, in create_dataset
    self[name] = dset
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/h5py/_hl/group.py", line 373, in __setitem__
    h5o.link(obj.id, self.id, name, lcpl=lcpl, lapl=self._lapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5o.pyx", line 202, in h5py.h5o.link
RuntimeError: Unable to create link (name already exists)
```

**Comment by padoremu** [comment]
I tried the workaround `model.save("rnn.h5", include_optimizer=False)` according to #27688, but this didn't help. I also tried to manually set different names for the cells with `GRUCell(10, name="cell" + str(i))`, which didn't help either.

**Comment by oanush** [comment]
@padoremu,
Can you try saving the model in tf format? using `model.save('rnn.tf',save_format='tf')`. It worked fine for me without any issue's. Kindly find the [gist of colab ](https://colab.sandbox.google.com/gist/oanush/84a2c6ef04084c18bbc699bc556c5acb/36093.ipynb#scrollTo=wJL_s5pKUetX)for your reference.Thanks!

**Comment by padoremu** [comment]
@oanush,
Thank you for your quick response. I tried with my 2.0.0-dev20190731 installation, but I am receiving an error `TypeError: wrapped_call() takes 1 positional argument but 2 were given`. Apparently I need a newer version of tensorflow.

I would love to install a current snapshot, but whatever I tried has failed so far:

1. Installing tf-nightly with pip results in installing tf-nightly-1.15.0.dev20190730, which is not as desired.

2. Installing a wheel package with pip fails with error `tf_nightly-2.2.0.dev20200121-cp36-cp36m-manylinux2010_x86_64.whl is not a supported wheel on this platform.`.

3. Source build fails eventually with error `Server terminated abruptly (error code: 14, error message: 'Socket closed'` (see https://github.com/tensorflow/models/issues/3647#issuecomment-577033825).

Is there any way to install e.g. a tf_nightly-2.2.0.dev20200121 (as seen on the gist of colab) on an Ubuntu 18.04, ideally without building from source?

**Comment by padoremu** [comment]
@oanush 
Have you tried loading the saved model? With my 2.0.0-dev20190731 installation, when adding the line

`model2 = tf.keras.models.load_model('rnn.tf')`

at the end, loading fails:

```
WARNING:tensorflow:Inconsistent references when loading the checkpoint into this object graph. Either the Trackable object references in the Python program have changed in an incompatible way, or the checkpoint was generated in an incompatible program.

Two checkpoint references resolved to different objects (ListWrapper([<tensorflow.python.keras.saving.saved_model.load.GRUCell object at 0x7f2c344becf8>, <tensorflow.python.keras.saving.saved_model.load.GRUCell object at 0x7f2c344bef28>]) and ListWrapper([ListWrapper([<tensorflow.python.keras.saving.saved_model.load.GRUCell object at 0x7f2c344becf8>, <tensorflow.python.keras.saving.saved_model.load.GRUCell object at 0x7f2c344bef28>]), ListWrapper([])])).
Traceback (most recent call last):
  File "test_rnn_gru.py", line 22, in <module>
    model2 = tf.keras.models.load_model('rnn.tf')
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/keras/saving/save.py", line 142, in load_model
    return saved_model_load.load(filepath, compile)
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/keras/saving/saved_model/load.py", line 86, in load
    model = tf_load.load_internal(path, loader_cls=KerasObjectLoader)
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/saved_model/load.py", line 541, in load_internal
    export_dir)
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/keras/saving/saved_model/load.py", line 102, in __init__
    super(KerasObjectLoader, self).__init__(*args, **kwargs)
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/saved_model/load.py", line 128, in __init__
    self._restore_checkpoint()
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/saved_model/load.py", line 281, in _restore_checkpoint
    load_status.assert_existing_objects_matched()
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/training/tracking/util.py", line 752, in assert_existing_objects_matched
    for trackable_object in self._graph_view.list_objects():
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/training/tracking/graph_view.py", line 430, in list_objects
    trackable_objects, _, _ = self.objects_ids_and_slot_variables()
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/training/tracking/graph_view.py", line 415, in objects_ids_and_slot_variables
    trackable_objects, path_to_root = self._breadth_first_traversal()
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/training/tracking/graph_view.py", line 199, in _breadth_first_traversal
    for name, dependency in self.list_dependencies(current_trackable):
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/training/tracking/graph_view.py", line 159, in list_dependencies
    return obj._checkpoint_dependencies
  File "/home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/training/tracking/data_structures.py", line 458, in _checkpoint_dependencies
    "automatically un-wrapped and subsequently ignored." % (self,)))
ValueError: Unable to save the object ListWrapper([ListWrapper([<tensorflow.python.keras.saving.saved_model.load.GRUCell object at 0x7f2c344becf8>, <tensorflow.python.keras.saving.saved_model.load.GRUCell object at 0x7f2c344bef28>]), ListWrapper([])]) (a list wrapper constructed to track trackable TensorFlow objects). A list element was replaced (__setitem__, __setslice__), deleted (__delitem__, __delslice__), or moved (sort). In order to support restoration on object creation, tracking is exclusively for append-only data structures.

If you don't need this list checkpointed, wrap it in a tf.contrib.checkpoint.NoDependency object; it will be automatically un-wrapped and subsequently ignored.
```

**Comment by oanush** [comment]
> @oanush,
> Thank you for your quick response. I tried with my 2.0.0-dev20190731 installation. It seems to work, but I am getting warnings:
> 
> ```
> WARNING:tensorflow:Entity <function Function._initialize_uninitialized_variables.<locals>.initialize_variables at 0x7fc8b0cd21e0> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Num'
> WARNING:tensorflow:Entity <function Function._initialize_uninitialized_variables.<locals>.initialize_variables at 0x7fc871a55048> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Num'
> WARNING:tensorflow:Entity <function Function._initialize_uninitialized_variables.<locals>.initialize_variables at 0x7fc8701499d8> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Num'
> WARNING:tensorflow:Entity <function Function._initialize_uninitialized_variables.<locals>.initialize_variables at 0x7fc871af60d0> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Num'
> WARNING:tensorflow:Entity <function Function._initialize_uninitialized_variables.<locals>.initialize_variables at 0x7fc84c67e1e0> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Num'
> WARNING:tensorflow:Entity <function Function._initialize_uninitialized_variables.<locals>.initialize_variables at 0x7fc84c401158> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Num'
> WARNING:tensorflow:Entity <function Function._initialize_uninitialized_variables.<locals>.initialize_variables at 0x7fc8146d6ea0> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Num'
> 2020-01-22 08:16:48.048918: W tensorflow/python/util/util.cc:288] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
> WARNING:tensorflow:Entity <function canonicalize_signatures.<locals>.signature_wrapper at 0x7fc8b0b65bf8> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
> WARNING:tensorflow:Entity <function Function._initialize_uninitialized_variables.<locals>.initialize_variables at 0x7fc8144619d8> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Num'
> WARNING:tensorflow:From /home/test/dev/tensorflow/tf2/lib/python3.6/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1784: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.
> ```
> 
> Apparently I need a newer tensorflow version, as in [gist of colab](https://colab.sandbox.google.com/gist/oanush/84a2c6ef04084c18bbc699bc556c5acb/36093.ipynb#scrollTo=wJL_s5pKUetX) with tf_nightly-2.2.0.dev20200121, most of the warnings do not occur. I would love to install a current snapshot, but whatever I tried has failed so far:
> 
> 1. Installing tf-nightly with pip results in installing tf-nightly-1.15.0.dev20190730, which is not as desired.
> 2. Installing a wheel package with pip fails with error `tf_nightly-2.2.0.dev20200121-cp36-cp36m-manylinux2010_x86_64.whl is not a supported wheel on this platform.`.
> 3. Source build fails eventually with error `Server terminated abruptly (error code: 14, error message: 'Socket closed'` (see [tensorflow/models#3647 (comment)](https://github.com/tensorflow/models/issues/3647#issuecomment-577033825)).
> 
> Is there any way to install e.g. a tf_nightly-2.2.0.dev20200121 (as seen on the gist of colab) on an Ubuntu 18.04, ideally without building from source?

@padoremu ,

Hello you can try installing nightly version with  [!pip install tf-nightly ](https://pypi.org/project/tf-nightly/), it should work fine!

**Comment by padoremu** [comment]
@oanush 
Unfortunately, that doesn't work for me at all, as in bash, the exclamation mark will substitute the command name with the last command in my history starting by that name - and this is somewhat arbitrary. What was your purpose of the exclamation mark?

If I create a new virtual environment and do `pip install tf-nightly`, it downloads `tf_nightly-1.15.0.dev20190730-cp36-cp36m-manylinux1_x86_64.whl`, which is why I end up with 1.15.0.dev20190730 being installed. By the way, if I download that wheel package manually and install it with pip, it works fine. However, installation of e.g. tf_nightly-2.2.0.dev20200121-cp36-cp36m-manylinux2010_x86_64.whl fails.

**Comment by padoremu** [comment]
@oanush Do you have any news? I tried to load the saved tf model in the colab, and it fails. So unfortunately, the workaround does not work, not even with the newest tensorflow version.

**Comment by jvishnuvardhan** [comment]
@padoremu In order to install `tf-nightly`, you can do this. At the command prompt, enter `pip3 install tf-nightly==2.2.0dev20200129`, then it will install `tf-nightly` that was released on 01/29 (today). I tried it on my Linux desktop and it works. Hope it works for you too. 

If this was resolved, please close the issue. Thanks!

**Comment by padoremu** [comment]
@jvishnuvardhan Thank you for your reply.

Regarding the model: For me, neither "h5" nor "tf" format work in the end. Saving in "tf" format works, but loading it doesn't  - can't see this issue being mentioned in the other ticket.

Regarding tf-nightly. Unfortunately, this doesn't work for me, I get the error:

```
Could not find a version that satisfies the requirement tf-nightly==2.2.0dev20200129 (from versions: 1.13.0.dev20190101, 1.13.0.dev20190102, 1.13.0.dev20190103, 1.13.0.dev20190104, 1.13.0.dev20190108, 1.13.0.dev20190109, 1.13.0.dev20190110, 1.13.0.dev20190111, 1.13.0.dev20190112, 1.13.0.dev20190113, 1.13.0.dev20190114, 1.13.0.dev20190115, 1.13.0.dev20190116, 1.13.0.dev20190117, 1.13.0.dev20190118, 1.13.0.dev20190119, 1.13.0.dev20190120, 1.13.0.dev20190121, 1.13.0.dev20190122, 1.13.0.dev20190123, 1.13.0.dev20190124, 1.13.0.dev20190125, 1.13.0.dev20190203, 1.13.0.dev20190204, 1.13.0.dev20190205, 1.13.0.dev20190206, 1.13.0.dev20190207, 1.13.0.dev20190208, 1.13.0.dev20190211, 1.13.0.dev20190212, 1.13.0.dev20190213, 1.13.0.dev20190214, 1.13.0.dev20190215, 1.13.0.dev20190216, 1.13.0.dev20190217, 1.13.0.dev20190218, 1.13.0.dev20190219, 1.13.0.dev20190220, 1.13.0.dev20190221, 1.13.0.dev20190222, 1.13.0.dev20190223, 1.13.0.dev20190224, 1.13.0.dev20190225, 1.13.0.dev20190226, 1.13.0.dev20190227, 1.13.0.dev20190228, 1.14.1.dev20190301, 1.14.1.dev20190302, 1.14.1.dev20190303, 1.14.1.dev20190304, 1.14.1.dev20190305, 1.14.1.dev20190306, 1.14.1.dev20190307, 1.14.1.dev20190308, 1.14.1.dev20190309, 1.14.1.dev20190310, 1.14.1.dev20190311, 1.14.1.dev20190312, 1.14.1.dev20190313, 1.14.1.dev20190314, 1.14.1.dev20190315, 1.14.1.dev20190316, 1.14.1.dev20190317, 1.14.1.dev20190318, 1.14.1.dev20190319, 1.14.1.dev20190320, 1.14.1.dev20190321, 1.14.1.dev20190322, 1.14.1.dev20190323, 1.14.1.dev20190324, 1.14.1.dev20190325, 1.14.1.dev20190326, 1.14.1.dev20190327, 1.14.1.dev20190329, 1.14.1.dev20190330, 1.14.1.dev20190331, 1.14.1.dev20190401, 1.14.1.dev20190402, 1.14.1.dev20190403, 1.14.1.dev20190404, 1.14.1.dev20190405, 1.14.1.dev20190406, 1.14.1.dev20190407, 1.14.1.dev20190408, 1.14.1.dev20190409, 1.14.1.dev20190410, 1.14.1.dev20190412, 1.14.1.dev20190413, 1.14.1.dev20190414, 1.14.1.dev20190415, 1.14.1.dev20190416, 1.14.1.dev20190417, 1.14.1.dev20190419, 1.14.1.dev20190420, 1.14.1.dev20190421, 1.14.1.dev20190422, 1.14.1.dev20190423, 1.14.1.dev20190424, 1.14.1.dev20190425, 1.14.1.dev20190426, 1.14.1.dev20190427, 1.14.1.dev20190428, 1.14.1.dev20190429, 1.14.1.dev20190430, 1.14.1.dev20190501, 1.14.1.dev20190502, 1.14.1.dev20190503, 1.14.1.dev20190504, 1.14.1.dev20190505, 1.14.1.dev20190506, 1.14.1.dev20190507, 1.14.1.dev20190508, 1.14.1.dev20190509, 1.14.1.dev20190510, 1.14.1.dev20190511, 1.14.1.dev20190512, 1.14.1.dev20190513, 1.14.1.dev20190514, 1.14.1.dev20190515, 1.14.1.dev20190516, 1.14.1.dev20190517, 1.14.1.dev20190518, 1.14.1.dev20190519, 1.14.1.dev20190520, 1.14.1.dev20190521, 1.14.1.dev20190522, 1.14.1.dev20190523, 1.14.1.dev20190524, 1.14.1.dev20190525, 1.14.1.dev20190526, 1.14.1.dev20190527, 1.14.1.dev20190528, 1.14.1.dev20190529, 1.14.1.dev20190530, 1.14.1.dev20190531, 1.14.1.dev20190601, 1.14.1.dev20190602, 1.14.1.dev20190603, 1.14.1.dev20190604, 1.14.1.dev20190605, 1.14.1.dev20190606, 1.14.1.dev20190607, 1.14.1.dev20190608, 1.14.1.dev20190611, 1.14.1.dev20190612, 1.14.1.dev20190613, 1.14.1.dev20190614, 1.14.1.dev20190615, 1.14.1.dev20190616, 1.14.1.dev20190617, 1.14.1.dev20190618, 1.14.1.dev20190619, 1.14.1.dev20190620, 1.14.1.dev20190621, 1.14.1.dev20190622, 1.14.1.dev20190623, 1.14.1.dev20190624, 1.14.1.dev20190625, 1.15.0.dev20190626, 1.15.0.dev20190627, 1.15.0.dev20190628, 1.15.0.dev20190703, 1.15.0.dev20190704, 1.15.0.dev20190705, 1.15.0.dev20190706, 1.15.0.dev20190707, 1.15.0.dev20190708, 1.15.0.dev20190709, 1.15.0.dev20190710, 1.15.0.dev20190711, 1.15.0.dev20190712, 1.15.0.dev20190713, 1.15.0.dev20190714, 1.15.0.dev20190715, 1.15.0.dev20190716, 1.15.0.dev20190717, 1.15.0.dev20190718, 1.15.0.dev20190723, 1.15.0.dev20190724, 1.15.0.dev20190725, 1.15.0.dev20190726, 1.15.0.dev20190727, 1.15.0.dev20190728, 1.15.0.dev20190729, 1.15.0.dev20190730)
No matching distribution found for tf-nightly==2.2.0dev20200129
```
What is your Linux and Python version? (for mine, see first posting here)

**Comment by jvishnuvardhan** [comment]
@padoremu Regarding the installation of `tf-nightly`, it is strange. You should be able to install with the command I provided. `pip3 install tf-nightly==2.2.0dev20200129`.

Can you please share the command you used? some time unknowingly we use pip instead of `pip3`.  
I have Python 3.6.8 and Linux 5.2.17-1rodete3-amd64 x86_64. 

I will check the saving issue and respond to you soon. Thanks!

**Comment by padoremu** [comment]
@jvishnuvardhan I had copy&pasted. Now also created a fresh venv:
```
test@test:~/dev/tensorflow$ python3 -m venv tf0
test@test:~/dev/tensorflow$ source ~/dev/tensorflow/tf0/bin/activate
(tf0) test@test:~/dev/tensorflow$ pip3 install tf-nightly==2.2.0dev20200129
```
Same result as already posted here. I have Python 3.6.9 and Linux 5.3.0-28-generic x86_64.

**Comment by jvishnuvardhan** [comment]
@padoremu Can you please follow the following steps to make virtual env.

```
Step 1: $:virtualenv --system-site-packages -p python3 ./venv_tfnightly

Step 2: source ./venv_tfnightly/bin/activate  

Step 3: pip3 install tf-nightly==2.2.0dev20200129

Step 4: $ python3

Step 5: >>> import tensorflow as tf
>>> tf.__version__  # it should print '2.2.0-dev20200129'

```
Thanks. Hope this helps.

**Comment by padoremu** [comment]
@jvishnuvardhan Thank you, that worked more or less, but after trying a full cleanup of my pip3 packages, because I got a warning due to some old tf packages (not in the virtualenv), something seems to be broken. So in the end I decided to use the nightly docker image, which works like a charm.

With the current nightly (2.2.0-dev20200202), loading the tf model works fine, which solves my problem, so I am closing the ticket. Thanks for the support!
